src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryTryGetValueAnalyzer.cs

```diff
@@ -0,0 +1,149 @@
+        private static readonly LocalizableString s_localizableTryGetValueDescription = CreateResource(nameof(MicrosoftNetCoreAnalyzersResources.PreferDictionaryTryGetValueDescription));
+
+        internal static readonly DiagnosticDescriptor ContainsKeyRule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableTryGetValueMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.BuildWarning,
+            s_localizableTryGetValueDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(ContainsKeyRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationContext)
+        {
+            var compilation = compilationContext.Compilation;
+
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericICollection1, out _))
+                return;
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIDictionary2, out var dictionaryType))
+                return;
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIEnumerable1, out _))
+                return;
+
+            compilationContext.RegisterOperationAction(context => OnOperationAction(context, dictionaryType), OperationKind.PropertyReference);
+        }
+
+        private void OnOperationAction(OperationAnalysisContext context, INamedTypeSymbol dictionaryType)
+        {
+            var propertyReference = (IPropertyReferenceOperation)context.Operation;
+
+            if (propertyReference.Parent is IAssignmentOperation
+                || !IsDictionaryAccess(propertyReference, dictionaryType)
+                || !TryGetParentConditionalOperation(propertyReference, out var conditionalOperation)
+                || !TryGetContainsKeyGuard(conditionalOperation, out var containsKeyInvocation))
+            {
+                return;
+            }
+
+            if (conditionalOperation!.WhenTrue is IBlockOperation blockOperation && DictionaryEntryIsModified(propertyReference, blockOperation))
+            {
+                return;
+            }
+
+            var additionalLocations = ImmutableArray.Create(propertyReference.Syntax.GetLocation());
+            context.ReportDiagnostic(Diagnostic.Create(ContainsKeyRule, containsKeyInvocation.Syntax.GetLocation(), additionalLocations));
```

Comment:
You may want to use the CreateDiagnostic extension method.

https://github.com/dotnet/roslyn-analyzers/blob/0919dbcf9bb03874aec10ec190c2f7f2397c10fb/src/Utilities/Compiler/Extensions/DiagnosticExtensions.cs

Reply:
Why? What advantages do I gain?